src/rules.jl

```diff
@@ -93,6 +95,133 @@ julia> hypot_pullback(1) == (NO_FIELDS, (x / hypot(x, y)), (y / hypot(x, y)))
 true
 ```
 
+When defining overloads they should be wrapped with the [`@rrule`](@ref) macro.
+
 See also: [`frule`](@ref), [`@scalar_rule`](@ref)
 """
 rrule(::Any, ::Vararg{Any}; kwargs...) = nothing
+
+"""
+    @frule(function ...)
+
+[`frule`](@ref) defining functions should be decorated with this macro.
+
+Example:
+```julia
+@frule function frule((Δself, Δargs...), ::typeof(foo), args...; kwargs...)
+    ...
+    return y, ∂y
+end
+```
+"""
+macro frule(ast)
+    return quote
+        $(esc(ast))
+        $_register_new_rule(frule, $(QuoteNode(ast)))
+    end
+end
+
+"""
+    @rrule(function ...)
+
+[`rrule`](@ref) defining functions should be decorated with this macro.
+
+Example:
+```julia
+@rrule function rrule(::typeof(foo), args...; kwargs...)
+    ...
+    return y, pullback
+end
+```
+"""
+macro rrule(ast)
+    return quote
+        $(esc(ast))
+        $_register_new_rule(rrule, $(QuoteNode(ast)))
+    end
+end
+
+"""
+    _rule_list(frule | rrule)
+
+Returns a list of all rules currently defined rules of the given kind.
+"""
+_rule_list
+_rule_list(::typeof(rrule)) = RRULES
+_rule_list(::typeof(frule)) = FRULES
+const SIGT = Type
```

Comment:
why this const? Is `SIGT` a more informative name?

Reply:
I thought I could do something betteer like 
`const SIGT = Type{NTuple{Type}}` 
which is what i am pretty sure it is, but I can't seem to get that to work

I do think it is a more informative name though